src/bin/svg2png/svg2png.cpp

```diff
@@ -25,180 +25,392 @@
 
-                 if (canvas->push(move(shape)) != tvg::Result::Success) return;
-            }
+            if (canvas->push(move(shape)) != tvg::Result::Success) return 1;
+        }
+
+        //Drawing
+        canvas->push(move(picture));
+        canvas->draw();
+        canvas->sync();
+
+        //Build Png
+        PngBuilder builder;
+        builder.build(dst, w, h, buffer, altColorspace);
 
-            //Drawing
-            canvas->push(move(picture));
-            canvas->draw();
-            canvas->sync();
+        //Clear canvas
+        canvas->clear(true);
 
-            //Build Png
-            PngBuilder builder;
-            builder.build(pngName.data(), width, height, buffer);
+        return 0;
+    }
 
-            //Terminate ThorVG Engine
-            tvg::Initializer::term(tvg::CanvasEngine::Sw);
+    void terminate()
+    {
+        //Terminate ThorVG Engine
+        tvg::Initializer::term(tvg::CanvasEngine::Sw);
 
-            free(buffer);
+        //Free buffer
+        free(buffer);
+    }
 
-        } else {
-            cout << "engine is not supported" << endl;
+private:
+    void _createCanvas()
+    {
+        //Canvas Engine
+        tvg::CanvasEngine tvgEngine = tvg::CanvasEngine::Sw;
+
+        //Threads Count
+        auto threads = thread::hardware_concurrency();
+
+        //Initialize ThorVG Engine
+        if (tvg::Initializer::init(tvgEngine, threads) != tvg::Result::Success) {
+            cout << "Error: Engine is not supported" << endl;
         }
+
+        //Create a Canvas
+        canvas = tvg::SwCanvas::gen();
+    }
+
+    void _createBuffer(int w, int h)
+    {
+        uint32_t size = w * h;
+        //Reuse old buffer if size is enough
+        if (buffer && buffer_size >= size) return;
+
+        //Alloc or realloc buffer
+        buffer = (uint32_t*) realloc(buffer, sizeof(uint32_t) * size);
+        buffer_size = size;
     }
 
-    int setup(int argc, char **argv, size_t *w, size_t *h)
+private:
+    unique_ptr<tvg::SwCanvas> canvas = nullptr;
+    uint32_t* buffer = nullptr;
+    uint32_t buffer_size = 0;
+};
+
+struct App
+{
+public:
+    int setup(int argc, char** argv)
     {
-        char *path{nullptr};
-
-        *w = *h = 0;
-
-        if (argc > 1) path = argv[1];
-        if (argc > 2) {
-            char tmp[20];
-            char *x = strstr(argv[2], "x");
-            if (x) {
-                snprintf(tmp, x - argv[2] + 1, "%s", argv[2]);
-                *w = atoi(tmp);
-                snprintf(tmp, sizeof(tmp), "%s", x + 1);
-                *h = atoi(tmp);
+        const char* path
+        { nullptr };
+
+        for (int i = 0, arg = 0; i < argc; i++) {
+            const char* p = argv[i];
+            if (*p == '-') {
+                //flags
+                if (p[1] == 'd') {
+                    //disable looking in subdirectories for wildcard
+                    flags |= FLAG_DISABLE_SUBDIRS;
+                } else if (p[1] == 'c') {
```

Comment:
Question: which case user need to select it? I don't see any point of it.

Reply:
For bugs testing. I think there may be some problems found using this flag.

Reply:
this has to be a useful tool as the svg2png, not a testing app. :(

Reply:
This is a user app, frankly, i don't agree on this, leave to @JSUYA s decision.